mirascope/mcp/client.py

```diff
@@ -0,0 +1,280 @@
+            else:
+                raise ValueError(f"Unsupported mime type: {mime_type}")
+
+    else:
+        raise ValueError(f"Unsupported content type: {type(content)}")
+
+
+def snake_to_pascal(snake: str) -> str:
+    return string.capwords(snake.replace("_", " ")).replace(" ", "")
+
+
+def build_object_type(
+    properties: dict[str, Any], required: list[str], name: str
+) -> type[BaseModel]:
+    fields = {}
+    for prop_name, prop_schema in properties.items():
+        class_name = snake_to_pascal(prop_name)
+        type_ = json_schema_to_python_type(prop_schema, class_name)
+        if prop_name in required:
+            fields[prop_name] = (
+                type_,
+                Field(..., description=prop_schema.get("description")),
+            )
+        else:
+            fields[prop_name] = (
+                type_ | None,
+                Field(None, description=prop_schema.get("description")),
+            )
+
+    return create_model(name, __config__=ConfigDict(extra="allow"), **fields)
+
+
+def json_schema_to_python_type(schema: dict[str, Any], name: str = "Model") -> Any:
+    """
+    Recursively convert a JSON Schema snippet into a Python type annotation or a dynamically generated pydantic model.
+    """
+    json_type = schema.get("type", "any")
+
+    if json_type == "string":
+        return str
+    elif json_type == "number":
+        return float
+    elif json_type == "integer":
+        return int
+    elif json_type == "boolean":
+        return bool
+    elif json_type == "array":
+        # Recursively determine the items type for arrays
+        items_schema = schema.get("items", {})
+        items_type = json_schema_to_python_type(items_schema)
+        return list[items_type]
+    elif json_type == "object":
+        # Recursively build a dynamic model for objects
+        properties = schema.get("properties", {})
+        required = schema.get("required", [])
+        return build_object_type(properties, required, name)
+    else:
+        # Default fallback
+        return Any
+
+
+def create_model_from_tool(tool: Tool) -> type[BaseModel]:
+    """
+    Create a dynamic pydantic model from the given Tool instance based on its inputSchema.
+    """
+    schema = tool.inputSchema
+    properties = schema.get("properties", {})
+    required_fields = schema.get("required", [])
+
+    fields = {}
+    for field_name, field_schema in properties.items():
+        field_type = json_schema_to_python_type(field_schema)
+
+        if field_name in required_fields:
+            default = Field(..., description=field_schema.get("description"))
+            annotation = field_type
+        else:
+            default = Field(None, description=field_schema.get("description"))
+            annotation = field_type | None
+
+        fields[field_name] = (annotation, default)
+
+    return create_model(
+        snake_to_pascal(tool.name), __config__=ConfigDict(extra="allow"), **fields
+    )
```

Comment:
The BaseModel is dynamically created from the JSON Schema here. Of course, this is a minimal implementation; it does not support json shcmea with complex nested structure of arguments types or references.
However,  I think this implementation is sufficient for the current situation.

Reply:
@brenkao do we still have our JSON schema to Pydantic model code somewhere? I feel like we wrote a pretty comprehensive version of this function that we could re-use here I'm just blanking on where that lives right now